sympy/solvers/solvers.py

```diff
@@ -3267,52 +3270,46 @@ def _Q(pow):
         return c.q
 
     # define the _take method that will determine whether a term is of interest
-    def _take(d, take_int_pow):
+    def _take(d):
         # return True if coefficient of any factor's exponent's den is not 1
         for pow in Mul.make_args(d):
-            if not (pow.is_Symbol or pow.is_Pow):
+            if not pow.is_Pow:
                 continue
-            b, e = pow.as_base_exp()
-            if not b.has(*syms):
+            if _Q(pow) == 1:
                 continue
-            if not take_int_pow and _Q(pow) == 1:
-                continue
-            free = pow.free_symbols
-            if free.intersection(syms):
+            if pow.free_symbols & syms:
                 return True
         return False
     _take = flags.setdefault('_take', _take)
 
+    if isinstance(eq, Eq):
+        eq = eq.rewrite(Add)
+        assert isinstance(eq, Expr)
+    elif not isinstance(eq, Expr):
+        return
```

Comment:
Maybe it would be better to only take Expr and passing an Eq should just be an error. All this does is convert the Eq to Expr but it would be better for the caller to do that I think.

I think we should not have `Eq.rewrite(Add)`. It's not really a rewrite since it turns into a semantically completely different kind of object. Also unless you use `evaluate=True` is it not just a very convoluted way of doing `eq.lhs - eq.rhs`?

Reply:
separate issues, but I thought the same about the first. I'm not sure it's worth breaking compatility to change it.

As for the `(Add)` rewrite, it creates an unevaluated expression. Not needed here but is important when trying to avoid cancellation of terms. I can rewrite here using the lhs/rhs with no issue, I believe, so then this will be one less place relying on it if it is ever removed.

Reply:
I wanted to stop Eq being used here when I made Relational not subclass Expr but I also wanted to minimise changes at the same time.